prov/efa/test/efa_unit_test_use_device_rdma.c

```diff
@@ -0,0 +1,118 @@
+#include "efa_unit_tests.h"
+
+
+void test_use_device_rdma(
+	const int env_val,
+	const int setopt_val,
+	const int expected_val,
+	const uint32_t api_version)
+{
+	int ret = 0;
+	struct fi_info *hints, *info;
+	struct fid_fabric *fabric = NULL;
+	struct fid_domain *domain = NULL;
+	struct fid_ep *ep = NULL;
+	struct rxr_ep *rxr_ep;
+	bool rdma_capable_hw;
+	char env_str[8];
+
+	if (env_val >= 0) {
+		snprintf(env_str,7,"%d",env_val);
+		setenv("FI_EFA_USE_DEVICE_RDMA",env_str, 1);
+	} else {
+		unsetenv("FI_EFA_USE_DEVICE_RDMA");
+	}
+
+	hints = efa_unit_test_alloc_hints(FI_EP_RDM);
+
+	ret = fi_getinfo(api_version, NULL, NULL, 0ULL, hints, &info);
+	assert_int_equal(ret, 0);
+
+	rdma_capable_hw = rxr_get_default_use_device_rdma();
+
+	if (expected_val && !rdma_capable_hw) {
+		/* cannot test USE_DEVICE_RDMA=1, hardware
+		   doesn't support it, and will abort() */
+		fi_freeinfo(info);
+		skip();
+		return;
+	}
+
+	ret = fi_fabric(info->fabric_attr, &fabric, NULL);
+	assert_int_equal(ret, 0);
+
+	ret = fi_domain(fabric, info, &domain, NULL);
+	assert_int_equal(ret, 0);
+
+	fi_endpoint(domain, info, &ep, NULL);
+	assert_int_equal(ret, 0);
+
+	if (setopt_val >= 0) {
+		bool b_val = (bool) setopt_val;
+		int ret_setopt;
+		ret_setopt = fi_setopt(&ep->fid, FI_OPT_ENDPOINT,
+			FI_OPT_EFA_USE_DEVICE_RDMA, &b_val, sizeof(bool));
+		if (FI_VERSION_LT(api_version, FI_VERSION(1,18))) {
+			assert_int_not_equal(ret_setopt, 0);
+		}
+		else if (expected_val != setopt_val) {
+			assert_int_not_equal(ret_setopt, 0);
+		}
+		else {
+			assert_int_equal(ret_setopt, 0);
+		}
+	}
+
+	rxr_ep = container_of(ep, struct rxr_ep,
+			base_ep.util_ep.ep_fid.fid);
+	assert_int_equal( rxr_ep->use_device_rdma, expected_val );
+
+	assert_int_equal(fi_close(&ep->fid), 0);
+	assert_int_equal(fi_close(&domain->fid), 0);
+	assert_int_equal(fi_close(&fabric->fid), 0);
+	fi_freeinfo(info);
+	unsetenv("FI_EFA_USE_DEVICE_RDMA");
```

Comment:
Hmmmm setting and resetting environment is tricky. If the test fails above this line then we end up with a dirty environment.

If we have to set/unsetenv, I think the correct way is to use the setup/teardown hook provided by cmocka.

Reply:
You are probably right.  Likely related, these tests can't come after the neuron tests, probably for a similar reason.

I think for these tests maybe we can just add the unsetenv() to teardown, because we want to setenv() to a variety of values

Reply:
moved the unset to the teardown hook.

Because setup is test-dependent, I'm still setting the value there.